### PR TITLE
make Service.init public to enable access outside of the HAP library

### LINKED ISSUE #45: Extending CharacteristicType
I am implementing a custom outlet accessory that measures both power and energy. For that reason `CharacteristicType` is extended with `.watt` and `. kiloWattHour`. Since there is no way to extend an enum this has to be done in the HAP library. It would however be desirable if you could avoid creating custom versions of the library, especially since I am using the proprietary characteristics and their UUIDs used by Elgato Eve, partly documented by [gomfunkel](https://gist.github.com/gomfunkel/b1a046d729757120907c). This way the energy sensor shows up in their app. 
 ```
extension Accessory {
    open class Energy: Accessory {
        public let energy = Service.Energy()
        
        public init(info: Service.Info) {
            super.init(info: info, type: .outlet, services: [energy])
        }
    }
}

extension Service {
    open class Energy: Service {
        public let on = GenericCharacteristic<Bool>(type: .on, value: false)
        public let inUse = GenericCharacteristic<Bool>(type: .outletInUse, value: true, permissions: [.read, .events])
        public let watt = GenericCharacteristic<Double>(type: .watt, value: 0, permissions: [.read, .events])
        public let kiloWattHour = GenericCharacteristic<Double>(type: .kiloWattHour, value: 0, permissions: [.read, .events])
        
        public init() {
            super.init(type: .outlet, characteristics: [on, inUse, watt, kiloWattHour])
        }
    }
}
```

Another problem is that `Accessory.init` seems to be internal and cannot be called outside the library, changing the permission opens a can of worms.

### PR DESCRIPTION
**Pull request by tgu**
Resolves #45 allowing for external access to Service.init and changing CharacteristicType to RawRepresentable struct allowing extensions and minimizing API change

### CODE REVIEW

File: Sources/HAP/Base/Constants.swift [hunk 1]
```diff
@@ -1,49 +1,78 @@
...
+    public init?(_ string: String) {
+        if let uuid = UUID(uuidString: string) {
+            self.rawValue = uuid
+        } else {
+            return nil
+        }
+    }
+
+    init(_ hex: UInt16) {
+        let string: String = "00000000\(String(hex, radix: 16))".suffix(8) + "-0000-1000-8000-0026BB765291"
```
[the remark this answers is not in the document]
Comment (tgu) [reply Sources/HAP/Base/Constants.swift]: Agree, maybe encapsulate in an enum.

[the remark this answers is not in the document]
Comment (Bouke) [reply Sources/HAP/Base/Constants.swift]: See also the HAP specification about this:

> 5.6.1 Service and Characteristic Types
> Service and Characteristic types are UUIDs as defined by RFC 4122. Apple-defined service and characteristic types are based on the following HAP Base UUID:
>
>     00000000-0000-1000-8000-0026BB765291
>
> When Apple-defined UUIDs are encoded as JSON strings, a short form must be used by including only the first 8 characters with leading zeros removed. When converting short form UUIDs back to full UUIDs, the process is reversed by prefixing the string with 0 or more '0' characters to expand it to 8 characters then "-0000-1000-8000-0026BB765291" is appended to form the full 36-character UUID.
>
> For example, the public.hap.service.accessory-information full UUID is "0000003E-0000-1000-8000-0026BB765291". When encoded as a JSON string, the first 8 characters are "0000003E" and with leading zeros removed it becomes "3E". Converting back to a full UUID would prefix "3E" with six '0' characters to get "0000003E" then append "-0000-1000-8000-0026BB765291" to form a full UUID of "0000003E-0000-1000-8000-0026BB765291". Other examples:
> 
>     "00000001-0000-1000-8000-0026BB765291" -> "1"
>     "00000F25-0000-1000-8000-0026BB765291" -> "F25"
>     "0000BBAB-0000-1000-8000-0026BB765291" -> "BBAB"
>     "010004FF-0000-1000-8000-0026BB765291" -> "10004FF"
>     "FF000000-0000-1000-8000-0026BB765291" -> "FF000000"
>
>
> Custom service and characteristic types must be 128-bit UUIDs as defined by RFC 4122. Custom types must be encoded as the full UUID string in JSON. Custom types must not use the HAP Base UUID.

**Comment by Bouke** [comment]
In 665aad2d445887e25c0da6c7fa783338be6730d6 I've made some changes that should make it possible to write a custom `Service`.

File: Sources/HAP/Base/Accessory.swift [hunk 2]
```diff
@@ -4,7 +4,7 @@ import Foundation
 // which is the data-type we wanted to use here. We can change it back
 // to UInt64 once the following commit has made it into a release:
 // https://github.com/apple/swift-corelibs-foundation/commit/64b67c91479390776c43a96bd31e4e85f106d5e1
-typealias InstanceID = Int
+public typealias InstanceID = Int
```
Comment (Bouke) [inline Sources/HAP/Base/Accessory.swift]: Could you revert this as well? Furthermore the PR is looking good.

**Comment by Bouke** [comment]
There are also a few test failures on Linux. I haven't looked into it much, but might have to do with JSON decoding. It's not as versatile as on macOS, on Linux the decodable type is defined by the JSON document. For example a JSON string cannot be deserialised into a number. That must be done by the user of the API.

**Comment by tgu** [comment]
Apple defined characteristic types are now in short for according to HAP 5.6.1

**Comment by tgu** [comment]
Trying with JSON coding with strings, maybe this will please linux. Completed the list of predefined types.

**Comment by Bouke** [comment]
PR merged, nice job!

**Pull request merged**